mono/mini/mini-runtime.c

```diff
@@ -996,6 +1009,7 @@ mono_thread_attach_cb (intptr_t tid, gpointer stack_start)
 	thread = mono_thread_info_current_unchecked ();
 	if (thread)
 		thread->jit_data = jit_tls;
+	mini_get_interp_callbacks ()->thread_attach_cb ();
```

Comment:
why is that needed? we usually lazily initialize `ThreadContext` in the interpreter.

Reply:
Not sure. I could try without.  I'm trying to follow what `MINT_MONO_JIT_ATTACH` used to do - which is a call to `set_context` after attaching to set `jit_tls->interp_context`.  Couldn't really tell from the code if that would always happen on its own otherwise.

Reply:
If I stub out the callback, `bug-58782-capture-and-throw.exe` fails with

```
* Assertion at interp/interp.c:5693, condition `context' not met
* thread #4, name = 'tid_1107', stop reason = signal SIGABRT
  * frame #0: 0x00007fff6be4ab66 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff6c015080 libsystem_pthread.dylib`pthread_kill + 333
    frame #2: 0x00007fff6bda61ae libsystem_c.dylib`abort + 127
    frame #3: 0x00000001002eeb40 mono`mono_log_write_logfile(log_domain=<unavailable>, level=<unavailable>, hdr=<unavailable>, message=<unavailable>) at mono-log-common.c:135 [opt]
    frame #4: 0x000000010030bf22 mono`monoeg_assertion_message [inlined] monoeg_g_logstr(log_domain=<unavailable>, log_level=G_LOG_LEVEL_ERROR, msg=<unavailable>) at goutput.c:117 [opt]
    frame #5: 0x000000010030bef8 mono`monoeg_assertion_message [inlined] monoeg_g_logv_nofree(log_domain=<unavailable>, log_level=G_LOG_LEVEL_ERROR, format=<unavailable>, args=0x0000003000000020) at goutput.c:128 [opt]
    frame #6: 0x000000010030bee4 mono`monoeg_assertion_message(format=<unavailable>) at goutput.c:163 [opt]
    frame #7: 0x0000000100111c85 mono`interp_set_resume_state(jit_tls=<unavailable>, ex=<unavailable>, ei=<unavailable>, interp_frame=<unavailable>, handler_ip=<unavailable>) at interp.c:5693 [opt]
    frame #8: 0x00000001000b400b mono`mono_handle_exception_internal(ctx=<unavailable>, obj=0x0000000102802030, resume=<unavailable>, out_ji=<unavailable>) at mini-exceptions.c:2568 [opt]
    frame #9: 0x00000001000b3359 mono`mono_handle_exception(ctx=<unavailable>, void_obj=<unavailable>) at mini-exceptions.c:2720 [opt]
    frame #10: 0x0000000100121d3a mono`interp_throw(context=0x00000001010015c0, ex=0x0000000102802030, frame=0x000070000eed5758, ip=0x0000000101814194, rethrow=0) at interp.c:734 [opt]
    frame #11: 0x0000000100117975 mono`interp_exec_method_full(frame=<unavailable>, context=0x00000001010015c0, start_with_ip=<unavailable>, filter_exception=0x0000000102802030, exit_at_finally=-1, base_frame=0x0000000000000000) at interp.c:4084 [opt]
    frame #12: 0x0000000100113bd8 mono`interp_exec_method_full(frame=<unavailable>, context=0x00000001010015c0, start_with_ip=<unavailable>, filter_exception=0x0000000000000000, exit_at_finally=-1, base_frame=0x0000000000000000) at interp.c:3012 [opt]
    frame #13: 0x0000000100113bd8 mono`interp_exec_method_full(frame=<unavailable>, context=0x00000001010015c0, start_with_ip=<unavailable>, filter_exception=0x00000001003e5030, exit_at_finally=-1, base_frame=0x0000000000000000) at interp.c:3012 [opt]
    frame #14: 0x0000000100112644 mono`interp_entry(data=<unavailable>) at interp.c:1749 [opt]
    frame #15: 0x00000001001240fc mono`interp_entry_static_0(rmethod=<unavailable>) at interp.c:2301 [opt]
    frame #16: 0x00000001016e9c99
    frame #17: 0x00007fff6c012661 libsystem_pthread.dylib`_pthread_body + 340
    frame #18: 0x00007fff6c01250d libsystem_pthread.dylib`_pthread_start + 377
    frame #19: 0x00007fff6c011bf9 libsystem_pthread.dylib`thread_start + 13
```

So it looks like without the attach callback we never set the interpreter callback on the natively-created thread.

Reply:
I'll take a look tomorrow

Reply:
pushed a fix

Reply:
Thanks! I'll update this PR. Could you backport to 2018-08 too

Reply:
already happened 🙂